cache/cache.go

```diff
@@ -0,0 +1,96 @@
+package cache
+
+import (
+	"container/list"
+)
+
+// Node represents a node eligible for caching.
+type Node interface {
+	GetKey() []byte
+}
+
+// Cache is an in-memory structure to persist nodes for quick access.
+type Cache interface {
+	// Adds node to cache. If full and had to remove the oldest element,
+	// returns the oldest, otherwise nil.
+	Add(node Node) Node
+
+	// Returns Node for the key, if exists. nil otherwise.
+	Get(key []byte) Node
+
+	// Has returns true if node with key exists in cache, false otherwise.
+	Has(key []byte) bool
+
+	// Remove removes node with key from cache. The removed node is returned.
+	// if not in cache, return nil.
+	Remove(key []byte) Node
+
+	// Len returns the cache length.
+	Len() int
+}
+
+// lruCache is an LRU cache implementation.
+type lruCache struct {
+	dict       map[string]*list.Element // FastNode cache.
+	cacheLimit int                      // FastNode cache size limit in elements.
+	ll         *list.List               // LRU queue of cache elements. Used for deletion.
+}
```

Comment:
I would prefer if we use https://github.com/hashicorp/golang-lru or https://github.com/hashicorp/golang-lru/tree/master/simplelru ? We are already using it in various places.

Reply:
I've looked into these options and, unfortunately, they do not allow for enough customization necessary for future work.

I'm planning to further improve the custom cache by adding the ability to have 2 kinds of limits:
- in terms of bytes
- in terms of the number of nodes

I have this work started here that I'm hoping to move to this repository: https://github.com/osmosis-labs/iavl/pull/40

More background about why I'm introducing this cache and its limits can be found here: https://github.com/osmosis-labs/osmosis/issues/1163

In short, we found it not intuitive configuring the limits in terms of nodes so we would like to provide the ability to do so in bytes only for fast cache

Reply:
Maybe let's add a short comment in the code, before someone else will come and propose to remove this and reuse other library.

Reply:
@odeke-em thanks for the suggestion. I've investigated using ristretto by swapping it for fast nodes but ran into issues. 

The major one is a ~17% increase in time/op. I suspect that is due to ristretto having overhead from concurrency guarantees which we might not really need.

Also, some randomized tests started failing for reasons I could not explain.

The current state of the attempt, summary, and benchmarks can be found here: https://github.com/cosmos/iavl/pull/508